Form and shape rework

CocytusDEDI opened the pull request:
# Form and shape rework
This rework closes #45.

## Done so far
- Shape rework basics
- Shape attribute in about section

## Still missing
- Form rework
- Shape changing component
- Energy consumption for changing natural shape size

CocytusDEDI committed:
Framework for form and shape rework

CocytusDEDI committed:
Shape basics refactor & code cleanup

CocytusDEDI committed:
Collision shape changes with form

The control required methods have also been added but not implemented yet

CocytusDEDI committed:
Another shape rework & shape section removal

CocytusDEDI committed:
Added shape change components & more

Codes docs where deleted and replaced with comments in the codes modules. The docs now point to it. Section code names where changed, shape codes where added and parameter types were moved into a code module

CocytusDEDI commented:
The basics have been done, but there is a bit of an issue now where the control needed for spells is unbalanced and it isn't communicated to the player why the spell is failing (it's failing due to not enough control).

AndycptSher commented on src/lib.rs:
```diff
@@ -661,13 +669,13 @@ impl Spell {
         let mut section: Option<u64> = None;
         while let Some(&bits) = instructions_iter.next() {
-            if section.is_some_and(|x| x == METADATA_SECTION) && !(READY_SECTION..=599).contains(&bits)  { // ignore all checks in metadata section
+            if section.is_some_and(|x| x == ABOUT_SECTION) && !(WHEN_CREATED_SECTION..=ABOUT_SECTION).contains(&bits)  {
```
### Word of caution
The refactoring of `599` to `ABOUT_SECTION` has potential to cause bugs when new sections are added in the future
This is applicable to all matches of sections

#### Possible remedy
creation of a new constant
> `pub const SECTIONS = 500..=599;`

and accompanying match pattern syntax (where required)
> ...
> `section if SECTIONS.constains(&section) => ...`
> ...

CocytusDEDI replied:
I think a better solution would be to redo the constants all together to put them into enums. This would prevent runtime errors as all the match statements that match the constants would no longer have catch all statements. Specifically in this case, it would prevent the future bugs when new sections are added and would also prevent the use of non-implemented numbers in the code. Doing so would also free up the namespace clutter and make it clearer what group the constants belong to.

CocytusDEDI replied:
Also, this branch is really starting to handle some things that it shouldn't be. So I won't be making the changes I suggested in this branch. I've opened #50.

CocytusDEDI committed:
Added charge_to_shape & generalised about parsing

CocytusDEDI commented:
`charge_to_shape` fixes the problem where if you used to much energy on a spell it would require tones of control, but the problem of balancing volume still exists. It currently requires a tone of energy to make shapes.

CocytusDEDI committed:
Volume balance change

The idea of using ln() functions for volume was scrapped as it was affecting shapes negatively

CocytusDEDI commented:
This code has a bug currently where control will go slightly negative after casting a spell when the player has no control. This becomes a problem in this branch though as when control is negative a spell is freed, and that spell is usually the spell you just cast since it's likely the largest one.

CocytusDEDI committed:
Focus change & issue fix

The issue was spells would be freed after casting a spell with no control left

CocytusDEDI commented:
The only thing that's still a bit bugged is that shapes use global directions instead of local ones to the player.

CocytusDEDI committed:
Forms now also use charge_to_shape & forms rotate to player
